File: 02_Sliding_Window/01_max_subarray_sum/solution.py
# GeeksforGeeks: Max Sum Subarray of Size K
# https://www.geeksforgeeks.org/problems/max-sum-subarray-of-size-k5313/1

class Solution:
    def maxSubarraySum(self, arr, k):
        if len(arr) < k:
            return 0

        window_sum = sum(arr[:k])
        max_sum = window_sum

        for i in range(k, len(arr)):
            window_sum += arr[i] - arr[i - k]
            max_sum = max(max_sum, window_sum)

        return max_sum

File: 02_Sliding_Window/01_max_subarray_sum/solution.cpp
class Solution {
  public:
    int maxSubarraySum(vector<int>& arr, int k) {
        int left = 0, sum = 0;
        int maxSum = INT_MIN;
        
        for (int right = 0; right < arr.size(); right++){
            sum += arr[right];
            
            if ((right - left + 1) == k){
                maxSum = max(maxSum, sum);
                sum -= arr[left];
                left++;
            }
        }
        return maxSum;
    }
};